pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,291 @@
+	ServiceName     types.NamespacedName
+	ListenPort      uint16
+	Add             bool
+	ResponseChannel chan bool
+}
+
+// ServiceEndpointsList: A list of endpoints for a service
+type ServiceEndpointsList struct {
+	ServiceName types.NamespacedName
+	endpoints   *sets.String
+}
+
+// serviceResponder: Contains net/http datastructures necessary for responding to each Service's health check on its aux nodePort
+type serviceResponder struct {
+	serviceName types.NamespacedName
+	listenPort  uint16
+	listener    *net.Listener
+	server      *http.Server
+}
+
+// ProxyHC: Handler structure for health check, endpoint add/delete and service listener add/delete requests
+type ProxyHC struct {
+	serviceEndpointsMap    map[types.NamespacedName]ServiceEndpointsList
+	serviceResponderMap    map[types.NamespacedName]serviceResponder
+	mutationRequestChannel chan *ProxyMutationRequest
+	hcRequestChannel       chan *ProxyHealthCheckRequest
+	listenerRequestChannel chan *ProxyListenerRequest
+	shutdownChannel        chan bool
+}
+
+// Utility function to send the required statusCode and error text to a http.ResponseWriter object
+func sendHealthCheckResponse(rw *http.ResponseWriter, statusCode int, error string) {
+	(*rw).Header().Set("Content-Type", "text/plain")
+	(*rw).WriteHeader(statusCode)
+	fmt.Fprint(*rw, error)
+}
+
+// Utility function to parse an incoming http request, construct and send a health check message to the main loop
+// Waits for responses from the main loop before responding to and closing the connection to the HTTP client.
+func parseHttpRequest(response *http.ResponseWriter, req *http.Request) (*ProxyHealthCheckRequest, chan *ProxyHealthCheckRequest, error) {
+	glog.V(3).Infof("Received Health Check on url %s", req.URL.String())
+	// Sanity check and parse the healthcheck URL
+	namespace, name, err := healthcheckparser.ParseURL(req.URL.String())
+	if err != nil {
+		glog.Info("Parse failure - cannot respond to malformed healthcheck URL")
+		return nil, nil, err
+	}
+	glog.V(4).Infof("Parsed Healthcheck as service %s/%s", namespace, name)
+
+	serviceName := types.NamespacedName{namespace, name}
+	responseChannel := make(chan *ProxyHealthCheckRequest, 1)
+	msg := &ProxyHealthCheckRequest{
+		ResponseChannel: &responseChannel,
+		ServiceName:     serviceName,
+		rw:              response,
+		req:             req,
+		Result:          false,
+	}
+	return msg, responseChannel, nil
+}
+
+// ServeHTTP: Interface callback method for net.Listener Handlers
+func (h *ProxyHC) ServeHTTP(response http.ResponseWriter, req *http.Request) {
+	// Grab the session guid from the URL and enqueue the request to the healtchecker
+	msg, responseChannel, err := parseHttpRequest(&response, req)
+	glog.V(3).Infof("Received HC Request Service %s from Cloud Load Balancer", msg.ServiceName)
+	if err != nil {
+		sendHealthCheckResponse(&response, http.StatusBadRequest, fmt.Sprintf("Parse error: %s", err))
+	}
+	h.hcRequestChannel <- msg
```

Comment:
I don't get the benefit of this buffered channel, please explain, or if you follow my suggestion below suggest just inlining  `handleHealthCheckRequest` here to make the code flow simpler.